pkg/documentlog/store/sql.go

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (C) 2020. Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package store
+
+import (
+	"bytes"
+	"errors"
+	"github.com/golang-migrate/migrate/v4"
+	"github.com/golang-migrate/migrate/v4/database/sqlite3"
+	bindata "github.com/golang-migrate/migrate/v4/source/go_bindata"
+	"github.com/jinzhu/gorm"
+	logging "github.com/nuts-foundation/nuts-network/logging"
+	"github.com/nuts-foundation/nuts-network/migrations"
+	"sync"
+
+	// import needed to enable the sqlite dialect for gorm
+	_ "github.com/jinzhu/gorm/dialects/sqlite"
+	"github.com/nuts-foundation/nuts-network/pkg/model"
+	"io"
+	"time"
+)
+
+const documentTable = "document"
+
+var documentColumns = []string{"hash", "type", "timestamp", "consistency_hash", "LENGTH(contents) as size"}
+
+func NewSQLDocumentStore(db *gorm.DB) (DocumentStore, error) {
+	docStore := &sqlDocumentStore{db: db, writeMutex: &sync.Mutex{}}
+	if err := docStore.migrate(); err != nil {
+		return nil, err
+	}
+	return docStore, nil
+}
+
+type sqlDocumentStore struct {
+	db         *gorm.DB
+	writeMutex *sync.Mutex
+}
+
+type sqlDocument struct {
+	Hash            string `gorm:"type:char(40);PRIMARY_KEY"`
+	ConsistencyHash string `gorm:"column:consistency_hash;type:char(40);PRIMARY_KEY"`
+	Type            string `gorm:"type:varchar(100)"`
+	Timestamp       int64  `gorm:"type:bigint"`
+	Size            int    `gorm:"-"`
+	Contents        []byte `gorm:"type:blob"`
+}
+
+func (d sqlDocument) storedDocument() StoredDocument {
```

Comment:
We now have `model.Document` and `model.DocumentDescriptor` in the `model` package which contains canonical datastructures used by all layers. Specific layers have their own representations which are internal for that layer (`p2p` has a protobuf one for serialization across the network, `store` has `sqlDocument` for storage using gorm).

Public datatypes and fields are documented, internal ones aren't (because they private to that layer). So from an architectural point of view, we only have to deal with Document and DocumentDescriptor which I don't feel need any more explanation than their comments in code.